scripts/check_copyright_notice.py

```diff
@@ -99,10 +120,114 @@
+]
+
+
+def get_modification_date(file: Path) -> datetime:
+    """Returns modification date for the file."""
+    date_string, _ = subprocess.Popen(  # pylint: disable=consider-using-with  # nosec
+        [str(GIT_PATH), "log", "-1", '--format="%ad"', "--", str(file)],
+        stdout=subprocess.PIPE,
+    ).communicate()
+    date_string_ = date_string.decode().strip()
+    if date_string_ == "":
+        return datetime.now()
+    return datetime.strptime(date_string_, '"%a %b %d %X %Y %z"')
+
+
+def get_year_data(match: re.Match, mixed: bool = False) -> Tuple[int, Optional[int]]:
+    """Get year data from match."""
+
+    if mixed:
+        year_string, *_ = match.groups()
+    else:
+        _, year_string, *_ = match.groups()
+
+    if "-" in year_string:
+        return (*map(int, year_string.split("-")),)  # type: ignore
+    return int(year_string), None
+
+
+def _validate_years(
+    file: Path,
+    allowed_start_years: Tuple[int, ...],
+    start_year: int,
+    end_year: int,
+    check_end_year: bool = True,
+) -> Tuple[bool, str]:
+    """
+    Given a file, check if the header stuff is in place.
+
+    Return True if the files has the encoding header and the copyright notice,
+    optionally prefixed by the shebang. Return False otherwise.
+
+    :param file: the file to check.
+    :param allowed_start_years: list of allowed start years
+    :param start_year: year when the file was created
+    :param end_year: year when the file was last modified
+    :param check_end_year: whether to validate end year or not
+    :return: True if the file is compliant with the checks, False otherwise.
+    """
+
+    modification_date = get_modification_date(file)
+
+    # Start year is not in allowed start years list
+    if start_year not in allowed_start_years:
+        return (
+            False,
+            f"Start year {start_year} is not in the list of allowed years; {allowed_start_years}.",
+        )
+
+    # Specified year is 2021 but the file has been last modified in another later year (missing -202x)
+    if end_year is not None and check_end_year:
+
+        if start_year > end_year:
+            return False, "End year should be greater then start year."
+
+        if end_year != modification_date.year:
+            return (
+                False,
+                f"End year does not match the last modification year. Header has: {end_year}; Last Modified: {modification_date.year}",
+            )
+
+    if end_year is None and modification_date.year > start_year:
+        return (
+            False,
+            f"Missing later year ({start_year}-20..)",
+        )
+
+    return True, ""
 
 
-def check_copyright(file: Path) -> bool:
+def _check_mixed(file: Path, content: str) -> Tuple[bool, str, str]:
+
+    check_status, message = _validate_years(
+        file,
+        START_YEARS_FETCHAI,
+        *get_year_data(FETCHAI_REGEX.search(content), True),  # type: ignore
+        check_end_year=False,
```

Comment:
We are not checking the last year for fetch.ai in mixed headers. Because most of the updates will be done by Valory so we will only validate the last year for Valory.